File: src/visual-graph/prune/resolve-ambiguous-queries.test.ts
import { describe, expect, test } from "vitest";

import type { ParsedRootQuery } from "../../cli/root-query/parsed-root-query.js";
import { ROOT_QUERY_KIND } from "../../cli/root-query/root-query-kind.js";
import { DIRECTION } from "../direction.js";
import type { VisualElement, VisualGraph, VisualNode } from "../model.js";
import { NODE_KIND } from "../node-kind.js";
import { VISUAL_ELEMENT_TYPE } from "../visual-element-type.js";
import { resolveAmbiguousQueries } from "./resolve-ambiguous-queries.js";

const variableNode = (name: string, line = 1): VisualNode => ({
  type: VISUAL_ELEMENT_TYPE.Node,
  id: `n-${name}-${String(line)}`,
  kind: NODE_KIND.Variable,
  name,
  line,
  endLine: null,
  isJsxElement: false,
  unused: false,
  declarationKind: null,
  initIsFunction: false,
});

const writeOpNode = (name: string, line = 1): VisualNode => ({
  type: VISUAL_ELEMENT_TYPE.Node,
  id: `w-${name}-${String(line)}`,
  kind: NODE_KIND.WriteOp,
  name,
  line,
  endLine: null,
  isJsxElement: false,
  unused: false,
  declarationKind: null,
});

const graphOf = (elements: VisualElement[]): VisualGraph => ({
  version: 1,
  source: { path: "f.ts", language: "ts" },
  direction: DIRECTION.TB,
  elements,
  edges: [],
  boundaryEdges: [],
  pruning: null,
});

const lineOrName = (raw: string, line: number): ParsedRootQuery => ({
  kind: ROOT_QUERY_KIND.LineOrName,
  line,
  name: raw,
  raw,
});

describe("resolveAmbiguousQueries", () => {
  test("returns input untouched when no LineOrName is present", () => {
    const graph = graphOf([variableNode("foo")]);
    const queries: ParsedRootQuery[] = [
      { kind: ROOT_QUERY_KIND.Line, line: 5, raw: "5" },
      { kind: ROOT_QUERY_KIND.Name, name: "foo", raw: "foo" },
    ];
    const result = resolveAmbiguousQueries(graph, queries);
    expect(result.resolved).toBe(queries);
    expect(result.resolutions).toEqual([]);
  });

  test("silently treats LineOrName as Line when no [Ll]\\d+ identifier exists", () => {
    const graph = graphOf([variableNode("foo"), variableNode("bar", 3)]);
    const result = resolveAmbiguousQueries(graph, [lineOrName("L12", 12)]);
    expect(result.resolved).toEqual([
      { kind: ROOT_QUERY_KIND.Line, line: 12, raw: "L12" },
    ]);
    expect(result.resolutions).toEqual([]);
  });

  test("resolves to Name when an exact match exists, with a resolution log", () => {
    const graph = graphOf([variableNode("L12", 7), variableNode("other", 9)]);
    const result = resolveAmbiguousQueries(graph, [lineOrName("L12", 12)]);
    expect(result.resolved).toEqual([
      { kind: ROOT_QUERY_KIND.Name, name: "L12", raw: "L12" },
    ]);
    expect(result.resolutions).toEqual([
      { raw: "L12", line: 12, name: "L12", resolvedAs: "name" },
    ]);
  });

  test("resolves to Line when other [Ll]<n> identifiers exist but no exact match", () => {
    const graph = graphOf([variableNode("l5"), variableNode("l99", 3)]);
    const result = resolveAmbiguousQueries(graph, [lineOrName("L12", 12)]);
    expect(result.resolved).toEqual([
      { kind: ROOT_QUERY_KIND.Line, line: 12, raw: "L12" },
    ]);
    expect(result.resolutions).toEqual([
      { raw: "L12", line: 12, name: "L12", resolvedAs: "line" },
    ]);
  });

  test("name lookup is case-sensitive", () => {
    // `l1` exists but the query is `L1`; matchableNames contains `l1`,
    // which still satisfies the [Ll]\d+ presence check, so we follow the
    // "exists but no exact match" path -> resolves to Line.
    const graph = graphOf([variableNode("l1")]);
    const result = resolveAmbiguousQueries(graph, [lineOrName("L1", 1)]);
    expect(result.resolved).toEqual([
      { kind: ROOT_QUERY_KIND.Line, line: 1, raw: "L1" },
    ]);
    expect(result.resolutions).toEqual([
      { raw: "L1", line: 1, name: "L1", resolvedAs: "line" },
    ]);
  });

  test("preserves order across a mixed array", () => {
    const graph = graphOf([variableNode("L12", 4), variableNode("l5", 6)]);
    const queries: ParsedRootQuery[] = [
      { kind: ROOT_QUERY_KIND.Line, line: 1, raw: "1" },
      lineOrName("L12", 12),
      { kind: ROOT_QUERY_KIND.Name, name: "x", raw: "x" },
      lineOrName("L99", 99),
    ];
    const result = resolveAmbiguousQueries(graph, queries);
    expect(result.resolved).toEqual([
      { kind: ROOT_QUERY_KIND.Line, line: 1, raw: "1" },
      { kind: ROOT_QUERY_KIND.Name, name: "L12", raw: "L12" },
      { kind: ROOT_QUERY_KIND.Name, name: "x", raw: "x" },
      { kind: ROOT_QUERY_KIND.Line, line: 99, raw: "L99" },
    ]);
    expect(result.resolutions).toEqual([
      { raw: "L12", line: 12, name: "L12", resolvedAs: "name" },
      { raw: "L99", line: 99, name: "L99", resolvedAs: "line" },
    ]);
  });

  test("emits one resolution entry per LineOrName occurrence", () => {
    const graph = graphOf([variableNode("L12")]);
    const result = resolveAmbiguousQueries(graph, [
      lineOrName("L12", 12),
      lineOrName("L12", 12),
    ]);
    expect(result.resolutions).toEqual([
      { raw: "L12", line: 12, name: "L12", resolvedAs: "name" },
      { raw: "L12", line: 12, name: "L12", resolvedAs: "name" },
    ]);
  });

  test("ignores name-query-excluded kinds when collecting matchable names", () => {
    // A WriteOp named `L12` is excluded from `-r <id>` matching, so the
    // resolver must not treat it as an exact match either. Without any
    // other [Ll]\d+ identifier, we fall to the silent Line path.
    const graph = graphOf([writeOpNode("L12"), variableNode("foo")]);
    const result = resolveAmbiguousQueries(graph, [lineOrName("L12", 12)]);
    expect(result.resolved).toEqual([
      { kind: ROOT_QUERY_KIND.Line, line: 12, raw: "L12" },
    ]);
    expect(result.resolutions).toEqual([]);
  });
});

File: crates/unsnarl-visual-graph/src/prune/resolve_ambiguous_queries_test.rs
use super::*;

use unsnarl_ir::SourceLine;

use crate::prune::test_helpers::{const_binding_node_with_end, graph_of, write_op_node};
use crate::visual_element::VisualElement;
use crate::visual_node::VisualNode;

fn variable_node(name: &str, line: u32) -> VisualNode {
    const_binding_node_with_end(&format!("n-{name}-{line}"), name, line, None)
}

fn writeop_named(name: &str, line: u32) -> VisualNode {
    write_op_node(&format!("w-{name}-{line}"), name, line)
}

fn line_or_name(raw: &str, line: u32) -> ParsedRootQuery {
    ParsedRootQuery::LineOrName {
        line: SourceLine(line),
        name: raw.to_string(),
        raw: raw.to_string(),
    }
}

fn line_query(line: u32, raw: &str) -> ParsedRootQuery {
    ParsedRootQuery::Line {
        line: SourceLine(line),
        raw: raw.to_string(),
    }
}

fn name_query(name: &str) -> ParsedRootQuery {
    ParsedRootQuery::Name {
        name: name.to_string(),
        raw: name.to_string(),
    }
}

#[test]
fn returns_input_untouched_when_no_line_or_name_is_present() {
    let g = graph_of(vec![VisualElement::Node(variable_node("foo", 1))], vec![]);
    let queries = vec![line_query(5, "5"), name_query("foo")];
    let result = resolve_ambiguous_queries(&g, &queries);
    assert_eq!(result.resolved, queries);
    assert!(result.resolutions.is_empty());
}

#[test]
fn silently_treats_line_or_name_as_line_when_no_l_prefix_identifier_exists() {
    let g = graph_of(
        vec![
            VisualElement::Node(variable_node("foo", 1)),
            VisualElement::Node(variable_node("bar", 3)),
        ],
        vec![],
    );
    let result = resolve_ambiguous_queries(&g, &[line_or_name("L12", 12)]);
    assert_eq!(result.resolved, vec![line_query(12, "L12")]);
    assert!(result.resolutions.is_empty());
}

#[test]
fn resolves_to_name_when_an_exact_match_exists_with_a_resolution_log() {
    let g = graph_of(
        vec![
            VisualElement::Node(variable_node("L12", 7)),
            VisualElement::Node(variable_node("other", 9)),
        ],
        vec![],
    );
    let result = resolve_ambiguous_queries(&g, &[line_or_name("L12", 12)]);
    assert_eq!(result.resolved, vec![name_query("L12")]);
    assert_eq!(
        result.resolutions,
        vec![RootQueryResolution {
            raw: "L12".to_string(),
            line: SourceLine(12),
            name: "L12".to_string(),
            resolved_as: ResolvedAs::Name,
        }]
    );
}

#[test]
fn resolves_to_line_when_other_l_n_identifiers_exist_but_no_exact_match() {
    let g = graph_of(
        vec![
            VisualElement::Node(variable_node("l5", 1)),
            VisualElement::Node(variable_node("l99", 3)),
        ],
        vec![],
    );
    let result = resolve_ambiguous_queries(&g, &[line_or_name("L12", 12)]);
    assert_eq!(result.resolved, vec![line_query(12, "L12")]);
    assert_eq!(
        result.resolutions,
        vec![RootQueryResolution {
            raw: "L12".to_string(),
            line: SourceLine(12),
            name: "L12".to_string(),
            resolved_as: ResolvedAs::Line,
        }]
    );
}

#[test]
fn name_lookup_is_case_sensitive() {
    let g = graph_of(vec![VisualElement::Node(variable_node("l1", 1))], vec![]);
    let result = resolve_ambiguous_queries(&g, &[line_or_name("L1", 1)]);
    assert_eq!(result.resolved, vec![line_query(1, "L1")]);
    assert_eq!(
        result.resolutions,
        vec![RootQueryResolution {
            raw: "L1".to_string(),
            line: SourceLine(1),
            name: "L1".to_string(),
            resolved_as: ResolvedAs::Line,
        }]
    );
}

#[test]
fn preserves_order_across_a_mixed_array() {
    let g = graph_of(
        vec![
            VisualElement::Node(variable_node("L12", 4)),
            VisualElement::Node(variable_node("l5", 6)),
        ],
        vec![],
    );
    let queries = vec![
        line_query(1, "1"),
        line_or_name("L12", 12),
        name_query("x"),
        line_or_name("L99", 99),
    ];
    let result = resolve_ambiguous_queries(&g, &queries);
    assert_eq!(
        result.resolved,
        vec![
            line_query(1, "1"),
            name_query("L12"),
            name_query("x"),
            line_query(99, "L99"),
        ]
    );
    assert_eq!(
        result.resolutions,
        vec![
            RootQueryResolution {
                raw: "L12".to_string(),
                line: SourceLine(12),
                name: "L12".to_string(),
                resolved_as: ResolvedAs::Name,
            },
            RootQueryResolution {
                raw: "L99".to_string(),
                line: SourceLine(99),
                name: "L99".to_string(),
                resolved_as: ResolvedAs::Line,
            },
        ]
    );
}

#[test]
fn emits_one_resolution_entry_per_line_or_name_occurrence() {
    let g = graph_of(vec![VisualElement::Node(variable_node("L12", 1))], vec![]);
    let result = resolve_ambiguous_queries(&g, &[line_or_name("L12", 12), line_or_name("L12", 12)]);
    assert_eq!(
        result.resolutions,
        vec![
            RootQueryResolution {
                raw: "L12".to_string(),
                line: SourceLine(12),
                name: "L12".to_string(),
                resolved_as: ResolvedAs::Name,
            },
            RootQueryResolution {
                raw: "L12".to_string(),
                line: SourceLine(12),
                name: "L12".to_string(),
                resolved_as: ResolvedAs::Name,
            },
        ]
    );
}

#[test]
fn ignores_name_query_excluded_kinds_when_collecting_matchable_names() {
    let g = graph_of(
        vec![
            VisualElement::Node(writeop_named("L12", 1)),
            VisualElement::Node(variable_node("foo", 3)),
        ],
        vec![],
    );
    let result = resolve_ambiguous_queries(&g, &[line_or_name("L12", 12)]);
    assert_eq!(result.resolved, vec![line_query(12, "L12")]);
    assert!(result.resolutions.is_empty());
}
